### ISSUE TITLE
Doctrine\ORM\ORMException: The EntityManager is closed

### ISSUE THREAD

**Issue by ahmadalli** [shlinkio/shlink#423]
Hi :)

my shlink instance (hosted on docker on a kubernetes cluster) becomes unstable after a while and returns 500 for all of the requests (except for 404s on `/` route) with this error on stdout:

```
Doctrine\ORM\ORMException: The EntityManager is closed. in /etc/shlink/vendor/doctrine/orm/lib/Doctrine/ORM/ORMException.php:144
Stack trace:
#0 /etc/shlink/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php(756): Doctrine\ORM\ORMException::entityManagerClosed()
#1 /etc/shlink/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php(589): Doctrine\ORM\EntityManager->errorIfClosed()
#2 /etc/shlink/module/Core/src/Service/VisitsTracker.php(42): Doctrine\ORM\EntityManager->persist(Object(Shlinkio\Shlink\Core\Entity\Visit))
#3 /etc/shlink/module/Core/src/Action/AbstractTrackingAction.php(64): Shlinkio\Shlink\Core\Service\VisitsTracker->track('dl', Object(Shlinkio\Shlink\Core\Model\Visitor))
#4 /etc/shlink/vendor/zendframework/zend-expressive/src/Middleware/LazyLoadingMiddleware.php(46): Shlinkio\Shlink\Core\Action\AbstractTrackingAction->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#5 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Expressive\Middleware\LazyLoadingMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#6 /etc/shlink/vendor/akrabat/ip-address-middleware/src/IpAddress.php(91): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#7 /etc/shlink/vendor/zendframework/zend-expressive/src/Middleware/LazyLoadingMiddleware.php(46): RKA\Middleware\IpAddress->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#8 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Expressive\Middleware\LazyLoadingMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#9 /etc/shlink/vendor/zendframework/zend-stratigility/src/MiddlewarePipe.php(83): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#10 /etc/shlink/vendor/zendframework/zend-expressive-router/src/Route.php(100): Zend\Stratigility\MiddlewarePipe->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#11 /etc/shlink/vendor/zendframework/zend-expressive-router/src/RouteResult.php(110): Zend\Expressive\Router\Route->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#12 /etc/shlink/vendor/zendframework/zend-expressive-router/src/Middleware/DispatchMiddleware.php(35): Zend\Expressive\Router\RouteResult->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#13 /etc/shlink/vendor/zendframework/zend-expressive/src/Middleware/LazyLoadingMiddleware.php(46): Zend\Expressive\Router\Middleware\DispatchMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#14 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Expressive\Middleware\LazyLoadingMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#15 /etc/shlink/vendor/zendframework/zend-stratigility/src/MiddlewarePipe.php(83): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#16 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Stratigility\MiddlewarePipe->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#17 /etc/shlink/vendor/zendframework/zend-stratigility/src/Middleware/PathMiddlewareDecorator.php(42): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#18 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Stratigility\Middleware\PathMiddlewareDecorator->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#19 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(47): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#20 /etc/shlink/vendor/zendframework/zend-expressive-router/src/Middleware/RouteMiddleware.php(54): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#21 /etc/shlink/vendor/zendframework/zend-expressive/src/Middleware/LazyLoadingMiddleware.php(46): Zend\Expressive\Router\Middleware\RouteMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#22 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Expressive\Middleware\LazyLoadingMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#23 /etc/shlink/vendor/zendframework/zend-stratigility/src/MiddlewarePipe.php(83): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#24 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Stratigility\MiddlewarePipe->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#25 /etc/shlink/vendor/zendframework/zend-stratigility/src/Middleware/PathMiddlewareDecorator.php(42): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#26 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Stratigility\Middleware\PathMiddlewareDecorator->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#27 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(47): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#28 /etc/shlink/module/Common/src/Middleware/CloseDbConnectionMiddleware.php(28): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#29 /etc/shlink/vendor/zendframework/zend-expressive/src/Middleware/LazyLoadingMiddleware.php(46): Shlinkio\Shlink\Common\Middleware\CloseDbConnectionMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#30 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Expressive\Middleware\LazyLoadingMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#31 /etc/shlink/vendor/zendframework/zend-expressive-helpers/src/ContentLengthMiddleware.php(31): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#32 /etc/shlink/vendor/zendframework/zend-expressive/src/Middleware/LazyLoadingMiddleware.php(46): Zend\Expressive\Helper\ContentLengthMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#33 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Expressive\Middleware\LazyLoadingMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#34 /etc/shlink/vendor/zendframework/zend-stratigility/src/Middleware/ErrorHandler.php(143): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#35 /etc/shlink/vendor/zendframework/zend-expressive/src/Middleware/LazyLoadingMiddleware.php(46): Zend\Stratigility\Middleware\ErrorHandler->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#36 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Expressive\Middleware\LazyLoadingMiddleware->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#37 /etc/shlink/vendor/zendframework/zend-stratigility/src/MiddlewarePipe.php(83): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#38 /etc/shlink/vendor/zendframework/zend-stratigility/src/Next.php(52): Zend\Stratigility\MiddlewarePipe->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\Next))
#39 /etc/shlink/vendor/zendframework/zend-stratigility/src/MiddlewarePipe.php(83): Zend\Stratigility\Next->handle(Object(Zend\Diactoros\ServerRequest))
#40 /etc/shlink/vendor/zendframework/zend-stratigility/src/MiddlewarePipe.php(72): Zend\Stratigility\MiddlewarePipe->process(Object(Zend\Diactoros\ServerRequest), Object(Zend\Stratigility\EmptyPipelineHandler))
#41 /etc/shlink/vendor/zendframework/zend-expressive-swoole/src/SwooleRequestHandlerRunner.php(244): Zend\Stratigility\MiddlewarePipe->handle(Object(Zend\Diactoros\ServerRequest))
#42 {main}
```

**Comment by ahmadalli** [comment]
I found out that this happens after I get a request with log parameters, e.g. 
`/sP57?utm_medium=native&utm_term=yn_item_00000&utm_campaign=adv_0000_0000000&utm_content=hmm.ok&utm_source=xxxxxxxxxxx&utm_yn=v2-OTM2ODk6MjQ0MzM6OTA3Mzo4Mg`

**Comment by ahmadalli** [comment]
it looks like it's like https://github.com/shlinkio/shlink/issues/253.
I've changed length of `visits`.`referer` to 4096, the issue was resolved. Since the referrer urls I was getting contained url encoded utf8 characters, the length would be more than 256 characters. I set it to 4096 just to be on the safe side

**Comment by ahmadalli** [comment]
Also I think database errors mustn't make system useless and unrecoverable and it's better to have a solution for it. 

Thank you for developing shlink btw. it's a useful product

**Comment by acelaya** [comment]
Hey @ahmadalli.

Thanks for reporting these two issues. None of them should be super hard to fix, so I'll add them to the roadmap.

**Comment by acelaya** [comment]
By the way, if you are using kubernetes, maybe you want to use [this endpoint](https://shlink.io/swagger-ui/index.html#/Monitoring/health) as the liveliness prove. That way, it will generate new pods if this error happens.

**Comment by ahmadalli** [comment]
@acelaya thanks for the endpoint :)

**Comment by acelaya** [comment]
Hey @ahmadalli 

Yesterday I released shlink 1.18, which includes a mitigation to the issue making shlink unusable after an error when using swoole.

I will address the field length issue next.

**Comment by acelaya** [comment]
@ahmadalli I have just released shlink 1.18.1, which includes a fix for this bug.

The size of the field has been increased from 256 to 1024, but also some logic has been put in place to crop the value before trying to persist it, in order to avoid a failure in the database.
